### ISSUE TITLE
woeusb hangs on 'install-grub' step

### ISSUE THREAD

**Issue by robinhood2014** [slacka/WoeUSB#251]
## Issue Reproduce Instructions
> 1. Launch WoeUSB by running `sudo woeusb --device Win10_1809Oct_English_x64.iso /dev/sdg --tgt-fs NTFS`
> 2. Wait until the program finishes `Copying files from source media...` and proceeds to `Installing GRUB bootloader for legacy PC booting support...`

## Expected Behavior
> `Installing for i386-pc platform.`
> `Install finished. No error reported.`

## Current Behavior
> `Installing for i386-pc platform.`
> The grub-install process hangs here forever, and I have to eventually press Ctrl-C to abort, leaving the USB disk unbootable.

### WoeUSB Source
> Installed from Ubuntu 18.04 software archive

### Information about the Operating System
> KDE neon User Edition 5.15, based on Ubuntu 18.04 "Bionic Beaver"

### Information about the Source Media
> "Windows 10" downloaded from <https://microsoft.com/en-us/software-download/Windows10ISO>

### Information about the Target Device
> PNY USB 2.0 FD, 16GB

**Comment by Lin-Buo-Ren** [comment]
It usually just _looks like_ forever as the install time will triple when using NTFS as the target filesystem, please try to reproduce again with the `--verbose` command-line option.

**Comment by cunhar** [comment]
what does _looks like forever_ look like in minutes? @Lin-Buo-Ren

**Comment by Lin-Buo-Ren** [comment]
@cunhar Last time I try it took about 0.5~1hr.

**Comment by cunhar** [comment]
@Lin-Buo-Ren --- can confirm it works --- just took about 0.5hr

thanks for the quick reply

**Comment by glaubitz** [comment]
Has this particular issue already been debugged? I am running into this problem as well and I would be happy to help to sort this issue out.

**Comment by WaxyMocha** [comment]
WoeUSB do not install grub by it self but calls external command to do it, and it looks like this command is taking ages to complete, so I don't think that there is much it can be done.

Code snippet:
```shell
echo_with_color green 'Installing GRUB bootloader for legacy PC booting support...'
"${command_grubinstall}" \
	--target=i386-pc \
	--boot-directory="${target_fs_mountpoint}" \
         --force "${target_device}"
```

**Comment by WaxyMocha** [comment]
But I'm not sure if grub is even necessary for creating bootable windows drive :thinking:

**Comment by glaubitz** [comment]
But ```grub-install``` does normally not hang that long. It either fails or completes successfully. So there must be something else that's causing the delay. I think I'll have a closer look at the source code.

**Comment by JonnyTech** [comment]
> I think I'll have a closer look at the source code.

@glaubitz did you find anything?

**Comment by JonnyTech** [comment]
I accidentally ran `woeusb` without the filesystem parameter:

```
sudo ./woeusb --device ./Win10_1903_V1_EnglishInternational_x64.iso /dev/sdb
```

The `Installing for i386-pc platform` stage completed quickly.

I then used the correct command line:

```
sudo ./woeusb --target-filesystem ntfs --device ./Win10_1903_V1_EnglishInternational_x64.iso /dev/sdb
```

But now the grub installation hung for a very, very, very long time.

Hopefully this may help @Lin-Buo-Ren to track down the issue.

**Comment by Lin-Buo-Ren** [comment]
That's actually expected behavior as by default WoeUSB creates an FAT32
filesystem which doesn't suffer from the poor write performance.

JonnyTech <notifications@github.com> 於 2019年6月25日 週二 上午1:30 寫道：

**Comment by Lin-Buo-Ren** [comment]
The NTFS:3G driver might be the cause, but I'm not really sure.

**Comment by samuelguesnier** [comment]
Same issue

**Comment by Azza-NZ** [comment]
> 
> 
> Same issue

https://win10.guru/usb-install-media-with-larger-than-4gb-wim-file/

I ended up doing this, super simple. 

Create a FAT32 and and NTFS partition on the the USB with  Gparted, Gnome Disks or KDE Partition Manager. Mount the ISO and copy all files to the NTFS and then only the files it says to the FAT32. I've tried it on 3 PC's and works great. I don't understand why WoeUSB and Rufus put there own EFI files on the USB.

**Comment by adamsb6** [comment]
Just investigated this in another issue, here: https://github.com/slacka/WoeUSB/issues/247#issuecomment-553635746

> It looks like this is grub2-install causing the hang. I strace'd it to find that it's sequentially copying a ton of files into the grub2 directory and is fsyncing after every write. See here for an example. In my case it's the fsync that's taking the most time, by far. Maybe slow fsync is a property of cheap USB flash devices.

> You can't fix the problem directly in WoeUSB, but you can hack around it. You could use an LD_PRELOAD trick to make fsync a no-op. In my case that'd probably speed things up by 10X. Then to make sure that everything is properly flushed, make a sync() call after grub2-install exits.

> You could also do these copies in parallel in grub2-install, but that would be outside the scope of the WoeUSB project.

I had not thought about the ntfs-3g driver referenced above, that could be the source of the slow fsyncs.

**Comment by samuelguesnier** [comment]
It finally worked after waiting more than one hour. However thanks for your advice @Azza-NZ

**Comment by bluedaze** [comment]
Alright, so I haven't spent much time on this, but I think that this might be solved by using grub-mkimge instead of grub-install...

After looking at the code and doing some research I have found that command_grubinstall invokes grub-install.

According to the docs for grub here at the [GNU GRUB Manual](http://manpages.ubuntu.com/manpages/trusty/man1/grub-mkimage.1.html) grub-install is a shell script which invokes grub-mkimage. According to the man page which I found here: [grub-mkimage](http://manpages.ubuntu.com/manpages/trusty/man1/grub-mkimage.1.htmll) there is an option for verbose output.

While this may not necessarily fix the issue I think it would be a great idea to try to use this instead as it will allow the user to know that there is progress being made, rather than wondering if the install is hanging.

I have zero experience with bash scripting but I will make an attempt at learning enough to see if this is a (temporarily) viable solution to the issue with the install hanging when I get home. I think there is probably a better solution but in the meantime, if someone wants to beat me to the issue... feel free!

**Comment by bluedaze** [comment]
Oh!

Also, I discovered that there is a flag for command_grubinstall which indicates that the drive is removable. "--removable"

Perhaps someone with more familiarity with the code might know why we don't use that flag? We are called WoeUSB, so I assume everyone here is using a USB... Maybe there is a edge case where this isn't the true?

**Comment by bluedaze** [comment]
Annnd, disregard my first comment above. You can actually invoke verbose mode with -v option in grub-install.

**Comment by nobodyinperson** [comment]
Currently running into this as well. Using `sudo strace -p $(pgrep -f grub-install)` you can see what the `grub-install` process actually does:

```c
openat(AT_FDCWD, "/usr/lib/grub/i386-pc/gcry_cast5.mod", O_RDONLY) = 4
openat(AT_FDCWD, "/media/woeusb_target_1579439133_10986/grub/i386-pc/gcry_cast5.mod", O_WRONLY|O_CREAT|O_TRUNC, 0644) = 5
read(4, "\177ELF\1\1\1\0\0\0\0\0\0\0\0\0\1\0\3\0\1\0\0\0\0\0\0\0\0\0\0\0"..., 1048576) = 16388
read(4, "", 1032188)                    = 0
write(5, "\177ELF\1\1\1\0\0\0\0\0\0\0\0\0\1\0\3\0\1\0\0\0\0\0\0\0\0\0\0\0"..., 16388) = 16388
read(4, "", 1048576)                    = 0
fsync(5
// ... same block, but for different file ...
// ...dito...
```

It seems to hang at the `fsync()` system call which might indicate that the either the filesystem driver or the device is insanely slow. The flash drive I am using can write with around 5 Mb/s, so it's most likely the filesystem driver (NTFS here).

Eventually, it succeeds.

**Comment by J0SH1X** [comment]
> Currently running into this as well. Using `sudo strace -p $(pgrep -f grub-install)` you can see what the `grub-install` process actually does:
> 
> ```c
> openat(AT_FDCWD, "/usr/lib/grub/i386-pc/gcry_cast5.mod", O_RDONLY) = 4
> openat(AT_FDCWD, "/media/woeusb_target_1579439133_10986/grub/i386-pc/gcry_cast5.mod", O_WRONLY|O_CREAT|O_TRUNC, 0644) = 5
> read(4, "\177ELF\1\1\1\0\0\0\0\0\0\0\0\0\1\0\3\0\1\0\0\0\0\0\0\0\0\0\0\0"..., 1048576) = 16388
> read(4, "", 1032188)                    = 0
> write(5, "\177ELF\1\1\1\0\0\0\0\0\0\0\0\0\1\0\3\0\1\0\0\0\0\0\0\0\0\0\0\0"..., 16388) = 16388
> read(4, "", 1048576)                    = 0
> fsync(5
> // ... same block, but for different file ...
> // ...dito...
> ```
> 
> It seems to hang at the `fsync()` system call which might indicate that the either the filesystem driver or the device is insanely slow. The flash drive I am using can write with around 5 Mb/s, so it's most likely the filesystem driver (NTFS here).
> 
> Eventually, it succeeds.

well also its copying insanely much mod files for grub which we wouldnt need at all , im kinda watching this progressing slowly and there are so many mod files for grub

**Comment by jasonbrown1965** [comment]
> It finally worked after waiting more than one hour. However thanks for your advice @Azza-NZ

Confirming above - nearly an hour for the woeusb to report a successful installation.

Similar timeframe waiting for the windows iso to boot into the opening keyboard and language options. I noped out at the 50 minute stage, then read comments about installs taking a very long time. Woops. Tried again, waiting both in Linux and Window for a total time of around three hours. That is, the Windows installer was also slows. Writing a win10 recovery drive as advised will take up to five hours on old gear, e.g. with only 4gb ram. Overall, win10 2004 version feels remarkably snappy for said old gear. Browsers still crawl to the start blocks, but once there tab up nicely.* 

On MX Linux, same memory, the **Ubuntu 18 version for woeusb worked**, the Ubuntu 20 did not. 

Be good to hear of other successful installs, or I just fluked it.

* am leaving out multiple user errors, because of mind-numbing stupidity, and no one needs to read about that.

**Comment by Yhozen** [comment]
It seems like this was never fixed :(

**Comment by jdavidls** [comment]
I propouse to change the string: 
'Info: Installing GRUB bootloader for legacy PC booting support...'
To: 
'Info: Installing GRUB bootloader for legacy PC booting support (this process can take a long time in some circumstances)...'

**Comment by BrianLeishman** [comment]
I get this hanging with the filesystem flag set to NTFS and without the flag at all, so even with a fat32 mine's hanging like this

**Comment by MiraiFoxx** [comment]
Is the problem still not resolved? Do I just need to wait or will they do something?

**Comment by Superduashwin** [comment]
It just took 1 hour , wait for some time or else do other work in you pc.... I also thought it would not finish but it go fineshed approximately 1 hour. ^_^

**Comment by olealgoritme** [comment]
This is insanely slow... Still not fixed.

**Comment by MiosrkFo** [comment]
bruh slow but worked its tooks like 2hrs or 3hrs

**Comment by ctr7pl** [comment]
Worked on Gentoo in around 10 seconds

**Comment by wheeskers** [comment]
Can confirm that this is not related to WoeUSB directly. This issue reproduces every time I try to install `grub` on NTFS partition of USB drive and the process takes about 30-40 minutes (it still finishes successfully though).

if ext4 is used then installation time is reduced dramatically. I did both partitioning and installation manually, no wrappers, USB tools, etc.

**Comment by bleszerd** [comment]
Just to keep updated: not fixed yet

**Comment by ivanjermakov** [comment]
Why installing grub step is event necessary? Make an option to skip it, not everyone needs it.

**Comment by bleszerd** [comment]
My problem was not having read the README, alternatively try using one of the versions described in the file!

![image](https://user-images.githubusercontent.com/77847606/154836515-2c2361ac-350e-4bd0-b01b-014cc68a1ec4.png)

**Comment by farseenmanekhan1232** [comment]
i was using woeusb-ng on arch linux , it takes around 40 - 50 mins to install ...

**Comment by tommy-mor** [comment]
Not recommended, but I did ctrl-c the process during grub install.. my usb still booted fine and I installed windows

**Comment by nahin-web** [comment]
It will take long if you have a slow USB device. Linux reports writing the data fast into the USB, but actually it takes longer for the write operation to sync with the memory cells of the device. You can notice it when you copy a large file into the thumb drive and when you chose 'eject disk' it will warn that the device is busy and after a long time the device can be safely removed.
